### ISSUE TITLE
GTDB-Tk v1.3.0, R95 and memory

### ISSUE THREAD

**Issue by davidealbanese** [Ecogenomics/GTDBTk#287]
Hi, I am observing memory peaks of ~300 GB with GTDB-Tk v1.3.0, R95, `--cpus 1` and `--pplacer_cpus 1` (`classify_wf` for 6 genomes) and not ~100 GB as reported in the [documentation](https://ecogenomics.github.io/GTDBTk/installing/index.html#hardware-requirements).

Any ideas?

Davide

**Comment by aaronmussig** [comment]
Hi Davide,

This due to how pplacer handles placing genomes. From my understanding of how [pplacer runs](https://github.com/matsen/pplacer/blob/f40a363e962cca7131f1f2d372262e0081ff1190/pplacer_src/pplacer_run.ml): following queue creation, the [main thread forks](https://github.com/matsen/pplacer/blob/f40a363e962cca7131f1f2d372262e0081ff1190/pplacer_src/pplacer_run.ml#L629
) `-j` times (aka `--pplacer_cpus``).

Behind the scenes, it is calling [Unix.fork](https://man7.org/linux/man-pages/man2/fork.2.html) which is "copy-on-write". The diagram [found here](https://www.geeksforgeeks.org/copy-on-write/) is a good representation of how it works. 

The main pplacer thread (that contains ~150 GB of cached values) will then be forked for each CPU, in your case 1. The forked child will have a mapping to the same 150 GB of memory, totalling 300 GB. 

Since pplacer doesn't write to the memory space in any of the children, it will effectively only be using the 150 GB. This is a known issue for some queueing systems that will stop your job for "using more memory than you requested".

In short, as long as you have ~150 GB of memory it should still run.

**Comment by davidealbanese** [comment]
Dear @aaronmussig, 

thank you for clarifying that! 

Since the common use of queueing system, cloud computing services (e.g. AWS) and workflow management systems (e.g. Nextflow), I suggest you to update the [documentation](https://ecogenomics.github.io/GTDBTk/installing/index.html#hardware-requirements) with this information.

Thanks again,
Davide

**Comment by aaronmussig** [comment]
Cheers, it will be in the next docs update:  https://github.com/Ecogenomics/GTDBTk/commit/53a702d5037f899a9d1875a588cefbbfdebc290e
